### ISSUE TITLE
Add feature to generate module documentation

### ISSUE THREAD

**Issue by jmtd** [cekit/cekit#268]
Consider this python script in @rcernich's fork of cct_module (look a bit familiar?):

https://github.com/rcernich/cct_module/blob/CLOUD-2477/generate_docs.py

Should we incorporate that functionality into cekit?

It would be a good thing to have it maintained somewhere properly for other module authors to use. The idea is it will generate a markdown (or asciidoc) file that could then be committed alongside a module source code, so that there was a good API documentation for the module in the source, browseable on GitHub.

Cekit as a place for this code: on the one hand this is not how cekit works for its other features (input: image YAML, output: Dockerfile etc); on the other hand, much of the machinery needed to implement this well (parsing module yaml, jinja integration, schemas etc) is in cekit already, and putting this functionality in would mean it was a one-stop-shop for image-or-module authors.

Thoughts @goldmann @dbecvarik ?

**Comment by rcernich** [comment]
In addition to documentation, which I think should be part of the tool suite (e.g. like javadoc), I think we should start to consider how we package, distribute and resolve modules.  I think it might be nice to package a set of modules in a zip, to ease distribution (e.g. jboss.container.core:cct_module:1.0.0.Final, borrowing from GAV).  Of course, this broadens the scope, but it also allows for using tags to release the contents of our modules as zips, directly from github.  Food for thought.

**Comment by dbecvarik** [comment]
I would love to see this to be part of Cekit. The only issue I see is where to place docs? What format? do we want to be able to run some local server to serve it? The cekit 2.1rc1 should be available today, can we have this early in 2.2?

**Comment by goldmann** [comment]
I would say 2.2 at earliest considering the fact that we need to have answers for above and probably more questions.

**Comment by jmtd** [comment]
Excellent, I can start cooking a PR.

**Comment by rcernich** [comment]
I think we follow the spec as defined by atomic.  This is a markdown file, so if we can format it well, vi/more/less should be good enough to view the text.  Eventually, we can look at tying this into a usage script.

**Comment by jmtd** [comment]
> I think we follow the spec as defined by atomic. This is a markdown file, so if we can format it well, vi/more/less should be good enough to view the text. Eventually, we can look at tying this into a usage script.

The problem is, the atomic spec doesn't mandate the markdown file, and I don't think it ever ends up in the images built by atomic-reactor: only the derivative help.1 (man/roff) file does.

**Comment by dbecvarik** [comment]
@jmtd I like the idea of multiple generators. we can have them as "plugins' similar to builders/generators. Can we write some spec for this so we can discuss it or raise quick PoC PR?

**Comment by rcernich** [comment]
I understand that none of this is mandated, but I think it provides a good enough starting point.  It would be our responsibility to add this to the image, which should be part of the effort here.  I think adding support for other formats can follow (i.e. the information should be the same, just formatted differently).

Let's start with some basic documentation (like what we have), get it integrated, then start iterating.

**Comment by dbecvarik** [comment]
@jmtd  this is implemented, right? can you close it if so?

**Comment by jmtd** [comment]
@dbecvarik no, we have implemented image documentation: effectively the aggregate of module information into one template that's shipped inside the image. This issue is about documenting module sources: i.e. generating a file that can be committed alongside module sources and documents their "API".

**Comment by dbecvarik** [comment]
I think to go with USAGE.md (or README.md as its processed by github by default) for module docs should be nicer. 
Polluting module/image descriptors with a lot of docs will render it hard to read. and if someone wants to go with something like literate style programming it can be done in README.md and then tangled to descriptors.

**Comment by jmtd** [comment]
I think to go with USAGE.md (or README.md as its processed by github by default) for module docs should be nicer.
> Polluting module/image descriptors with a lot of docs will render it hard to read. 

The plan wasn't to add this stuff into descriptors - but to consume the descriptors and generate a document file (could be USAGE.md, sure) that tabulates the environment variable definitions, etc.

This might be easier to reason about in a PR once I've cooked the PoC.
